lib/msal-common/src/client/ClientCredentialClient.ts

```diff
@@ -0,0 +1,86 @@
+import { ServerAuthorizationTokenResponse } from "../response/ServerAuthorizationTokenResponse";
+import { RequestParameterBuilder } from "../request/RequestParameterBuilder";
+import { ScopeSet } from "../request/ScopeSet";
+import { GrantType } from "../utils/Constants";
+import { ResponseHandler } from "../response/ResponseHandler";
+import { AuthenticationResult } from "../response/AuthenticationResult";
+import { ClientCredentialRequest } from "../request/ClientCredentialRequest";
+
+/**
+ * OAuth2.0 client credential grant
+ */
+export class ClientCredentialClient extends BaseClient {
+
+    constructor(configuration: ClientConfiguration) {
+        super(configuration);
+    }
+
+    public async acquireToken(request: ClientCredentialRequest): Promise<AuthenticationResult>{
+        const response = await this.executeTokenRequest(request, this.authority);
+        // TODO delete this
+        console.log(response.body);
+        const responseHandler = new ResponseHandler(
+            this.config.authOptions.clientId,
+            this.cacheManager,
+            this.cryptoUtils,
+            this.logger
+        );
+
+        responseHandler.validateTokenResponse(response.body);
+        const tokenResponse = responseHandler.handleServerTokenResponse(
+            response.body,
+            this.authority,
+            null,
+            null,
+            request.scopes
+        );
+
+        return tokenResponse;
+    }
+
+    private async executeTokenRequest(request: ClientCredentialRequest, authority: Authority)
+        : Promise<NetworkResponse<ServerAuthorizationTokenResponse>> {
+
+        const requestBody = this.createTokenRequestBody(request);
+        const headers: Map<string, string> = this.createDefaultTokenRequestHeaders();
+
+        return this.executePostToTokenEndpoint(authority.tokenEndpoint, requestBody, headers);
+    }
+
+    private createTokenRequestBody(request: ClientCredentialRequest): string {
+        const parameterBuilder = new RequestParameterBuilder();
+
+        parameterBuilder.addClientId(this.config.authOptions.clientId);
+
+        const scopeSet = new ScopeSet(request.scopes || []);
+        parameterBuilder.addScopes(scopeSet);
+        
+        parameterBuilder.addGrantType(GrantType.CLIENT_CREDENTIALS_GRANT);
+
+        parameterBuilder.addClientInfo();
+
+        const correlationId = request.correlationId || this.config.cryptoInterface.createNewGuid();
+        parameterBuilder.addCorrelationId(correlationId);
+
+        if (this.config.clientCredentials.clientSecret) {
+            parameterBuilder.addClientSecret(this.config.clientCredentials.clientSecret);
+        }
+
+        if (this.config.clientCredentials.clientAssertion) {
+            const clientAssertion = this.config.clientCredentials.clientAssertion;
+            parameterBuilder.addClientAssertion(clientAssertion.assertion);
+            parameterBuilder.addClientAssertionType(clientAssertion.assertionType);
+        }
```

Comment:
Do we have empty/null checks on the clientCredentials params?

Reply:
Yes it is in ConfidentialClientApplication.ts